sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/TableFileCatalog.scala

```diff
@@ -55,10 +55,16 @@ class TableFileCatalog(
 
   override def listFiles(filters: Seq[Expression]): Seq[Partition] = partitionSchema match {
     case Some(partitionSchema) =>
-      externalCatalog.listPartitionsByFilter(db, table, filters).flatMap {
+      val catalogTablePartitions = externalCatalog.listPartitionsByFilter(db, table, filters)
```

Comment:
Actually, this breaks inputFiles as well when there are mixed partition paths as in "SPARK-15248: explicitly added partitions should be readable".

I'm going to look into this more.